core/server/common/src/main/java/alluxio/master/journal/AsyncJournalWriter.java

```diff
@@ -99,49 +162,195 @@ public long appendEntry(JournalEntry entry) {
+            // No more entries in the queue. Break write session.
             break;
           }
           mJournalWriter.write(entry);
           // Remove the head entry, after the entry was successfully written.
           mQueue.poll();
-          writeCounter = mWriteCounter.incrementAndGet();
+          mWriteCounter++;
 
-          if (writeCounter >= targetCounter) {
-            if ((System.nanoTime() - startTime) >= mFlushBatchTimeNs) {
-              // This thread has been writing to the journal for enough time. Break out of the
-              // infinite for-loop.
-              break;
+          if (((System.nanoTime() - startTime) >= mFlushBatchTimeNs) && !mStopFlushing) {
+            // This thread has been writing to the journal for enough time. Break out of the
+            // infinite while-loop.
+            break;
+          }
+        }
+
+        // Either written new entries or previous flush had been failed.
+        if (mFlushCounter.get() < mWriteCounter) {
+          mJournalWriter.flush();
+          mFlushCounter.set(mWriteCounter);
+        }
+
+        /**
+         * Notify tickets that have been served to wake up.
+         */
+        try (LockResource lr = new LockResource(mTicketLock)) {
+          ListIterator<FlushTicket> ticketIterator = mTicketList.listIterator();
+          while (ticketIterator.hasNext()) {
+            FlushTicket ticket = ticketIterator.next();
+            if (ticket.getTargetCounter() <= mFlushCounter.get()) {
+              ticket.setCompleted();
+              ticketIterator.remove();
             }
           }
         }
+      } catch (IOException | JournalClosedException exc) {
+        /**
+         * Release only tickets that have been flushed. Fail the rest.
+         */
+        try (LockResource lr = new LockResource(mTicketLock)) {
+          ListIterator<FlushTicket> ticketIterator = mTicketList.listIterator();
+          while (ticketIterator.hasNext()) {
+            FlushTicket ticket = ticketIterator.next();
+            ticketIterator.remove();
+            if (ticket.getTargetCounter() < mFlushCounter.get()) {
+              ticket.setCompleted();
+            } else {
+              ticket.setError(exc);
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Submits a ticket to flush thread and waits until ticket is served.
+   *
+   * If the specified counter is already flushed, this is essentially a no-op.
+   *
+   * @param targetCounter the counter to flush
+   */
+  @SuppressWarnings("Duplicates")
+  public void flush(final long targetCounter) throws IOException, JournalClosedException {
+    // Return if flushed.
+    if (targetCounter <= mFlushCounter.get()) {
+      return;
+    }
+
+    /**
+     * Submit the ticket for flush thread to process.
+     */
+    FlushTicket ticket = new FlushTicket(targetCounter);
+    try (LockResource lr = new LockResource(mTicketLock)) {
+      mTicketList.add(ticket);
```

Comment:
If the journal is not flushable for a long period of time, will we add an unbounded number of tickets to the list?

Reply:
Those later tickets will end up waking up the flush thread and cause it to retry the flush eventually. And that will either complete those tickets or fail them. So even though there is no hard bound on the ticket list, it's not expected to grow beyond active thread count.

Reply:
Got it, thanks.